Monorepo docs (#684)
packit/packit.dev: content/docs/Configuration/jobs.md

---
title: Packit Service jobs configuration 
date: 2023-07-03
weight: 1
---

# Packit Service jobs configuration

Packit Service doesn't have any web interface, but its behaviour can be configured via the config file 
generally described [here](/docs/configuration).

The tasks Packit Service should do should be defined in the `jobs` section. The section is a list of dictionaries:
```yaml
jobs:
- {key: value}
- {}
```

When there is an action (pull request, commit, etc.) in the upstream repository, Packit Service takes the config 
file from the relevant git ref (release tag for release, PR head for pull requests, related dist-git commit for downstream jobs).
Then, it tries to find a matching job in the `jobs` sections for the particular event (e.g. for pull request, jobs with
`pull_request` trigger are matched).

> If there is no `jobs` section in the configuration file, jobs default to:

Reviewer: Off-topic, but would be interesting to check if anyone is actually using the default value.

Also, we can add  koji and bodhi into the default .

Author: good suggestions! I will check and we can discuss

Author: Looks like not many users:
```
Affected repos (3): 
https://github.com/varlink/libvarlink
https://github.com/skupperproject/skupper-router
https://github.com/candlepin/virt-who
```

---
title: Packit Service jobs configuration 
date: 2023-07-03
weight: 1
---

# Packit Service jobs configuration

> Packit Service doesn't have any web interface to configure it,
> so the only way to change its behaviour is via the config file generally described [here](/docs/configuration).

Reviewer: ```suggestion
Packit Service doesn't have any web interface to configure it,
but you can change its behaviour via the config file generally described [here](/docs/configuration).
```

It sounded a bit negative to me.

When you open a pull request against your upstream repository, packit service
picks up configuration file from your pull request, not from the branch against
the PR is opened. This way, you can polish your .packit.yaml and see the

> results right away. (for more info, please see [packit-service#48](https://github.com/packit/packit-service/issues/48))

Reviewer: I would mention that it works like that in general -- that it takes the config from the relevant git ref (release tag for release, PR head for PRs, related dist-git commit for downstream jobs...).

Author: I rewrote this section that was written few years ago, PTAL

  trigger: pull_request
  targets: [fedora-stable]

- job: tests
  trigger: pull_request
  targets: [fedora-stable]

- job: propose_downstream
  trigger: release
  dist_git_branches:
    - fedora-all
```

If you do not want to use the jobs then the `jobs` section in the configuration file should be empty:
```yaml
jobs: []

```

Packit configuration supports [YAML Merge Key syntax](https://yaml.org/type/merge.html), which can be used to reduce duplication of configuration. Please see the example:
```yaml
# before
jobs:
- job: copr_build
  trigger: pull_request
  targets:
  - centos-stream-8-x86_64
  - centos-stream-9-x86_64
  - fedora-all

- job: copr_build
  trigger: commit
  branch: main
  targets:
  - centos-stream-8-x86_64
  - centos-stream-9-x86_64
  - fedora-all

# after
jobs:
- &copr
  job: copr_build
  trigger: pull_request
  targets:
  - centos-stream-8-x86_64
  - centos-stream-9-x86_64
  - fedora-all

- <<: *copr
  trigger: commit
  branch: main
```

## Configuration keys
#### job
(*str*) name of the job (you can imagine this as a CLI command).

#### trigger
(*str*) what is the trigger for the job? Every job only supports a specific set of triggers.

#### packages

> (*list*) Optional, list of package object names (when using monorepositories). If a job object has no such key, all packages of a monorepo should be handled.

Reviewer: Maybe a link to the top level `packages` key description?
